Подключение Vuex. Работа с глобальным состоянием приложения

efiand opened the pull request:
Пытался осмыслить то, что не надо заменять все пропсы на хранилище. Показалось разумным ограничить работу с хранилищем корзины и заказов на уровне вьюх, чтобы не тянуть через цепочку лэйаутов (а пользователя на уровне лэйаута). Для билдера показалось правильным не делать стор, так как там все это в пределах лишь одной страницы и уже разделено по ответственности между компонентами.
Также, помня о неоднозначности миксинов и том, что у меня там какая-то нездоровая конструкция с ценами получилась, решил вынести логику формирования и отображения цены в отдельный компонент (найдя решение, как унифицировать вывод цены за пиццу и стоимости всего заказа).

P.S. Старался осмысленно поделить на коммиты, но не удалось выделить компонент цены в отдельный коммит, так как уже много всего было намешано, творческий процесс шел стихийно.

---
:mortar_board: [Подключение Vuex. Работа с глобальным состоянием приложения](https://up.htmlacademy.ru/vue/1/user/367335/tasks/5)

efiand committed:
Убрал миксин content-prop
Остальные не стал, так как много связей с компонентами

efiand committed:
Вынос текущего заказа в хранилище и компонент цены

efiand committed:
Вынос списка заказов в хранилище

efiand committed:
Опечатка

efiand committed:
Вынос временного состояния пользователя

efiand committed:
Унификация кнопки закрытия через компонент

efiand committed:
Окно "Спасибо за заказ"

wspwebben commented on src/frontend/src/store/modules/auth.js:
```diff
@@ -0,0 +1,18 @@
+export default {
+  namespaced: true,
+  state: {
+    currentUser: {},
```
тут больше подойдет null

efiand replied:
А как тогда лучше сделать при передаче через пропс? Там сейчас обязательный объект ждет. Корректно ли изменить на `default: () => ({}),` ?

wspwebben replied:
нет, default срабатывает только для undefined, поэтому лучше использовать `default: null`

мне в целом не очень нравится идея использовать пустой объект в качестве пользователя, потому что при сравнении он тоже будет истинным

efiand replied:
В общем, когда я делаю currentUser: null, выскакивает ошибка, что пришел неверный тип пропса. А если при этом required заменить на default, то такой ошибки нет. То есть, дело было не в том, что я добавил default, а в том, что убрал required?

wspwebben commented on src/frontend/src/store/modules/cart.js:
```diff
@@ -0,0 +1,56 @@
+    [UPDATE_PIZZA]({ currentOrder }, { pizza, index }) {
+      const { counter } = currentOrder.pizzas[index];
+
+      currentOrder.pizzas[index] = {
```
во вьюксе действуют такие же правила реактивности, как и в компонентах, т.е. обновление по индексу не реактивно

нужно использовать Vue.set или пересобрать весь массив с помощью .map

wspwebben commented on src/frontend/src/store/modules/cart.js:
```diff
@@ -0,0 +1,56 @@
+    currentOrder: createOrder(),
+  },
+  mutations: {
+    [ADD_PIZZA]({ currentOrder }, pizza) {
```
стейт лучше не деструктурировать, чтобы переменная всегда ссылалась на правильную ссылку

wspwebben commented on src/frontend/src/store/modules/orders.js:
```diff
@@ -0,0 +1,16 @@
+      orders.push(newOrder);
+    },
+    [DELETE_ORDER]({ orders }, order) {
+      orders.splice(orders.indexOf(order), 1);
```
что будет, если переданного элемента не окажется в массиве?

efiand replied:
Беда будет ((
чет не сообразил, все проверки шли из интерфейса, где наличие кнопки удаления собственно наличием элемента и обусловлено.

efiand committed:
Работа над ошибками
Пока заменил обязательность пропса user на генерацию пустого объекта. Возможно, в теме по аутентификации будет понятно, как вообще лучше с пользователем поступать.

wspwebben commented:
👍

wspwebben commented on src/frontend/src/layouts/AppLayoutHeader.vue:
```diff
@@ -49,7 +49,7 @@ export default {
     user: {
       type: Object,
-      required: true,
+      default: () => ({}),
```
если не передать поле, то мы получим пустой объект, которые пройдет проверку в v-if, но полей для вывода у него не будет

сейчас в этом случае мы просто ничего не выведем, кроме лишней верстки, но если добавятся вложенные поля, то при попытке получить какой-нибудь `user.image.small`, то мы упадем с ошибкой

efiand replied:
Если убрать эту проверку, то все работает, я ее добавил на замену required: true, по рекомендации учебника. Проблема начинается, если вернуть `required: true` - `Invalid prop: type check failed for prop "user". Expected Object, got Null `. Как же быть: в данном случае достаточно только указать тип, или надо написать кастомный валидатор?

wspwebben replied:
нужно указать, что в пропсе может быть 2 типа: `type: [Object, null]`

efiand replied:
Спасибо! Это я упустил по программе. Поправлю.

efiand replied:
![изображение](https://user-images.githubusercontent.com/26799424/173032180-c6ce5543-fba8-480a-a1f0-0bb7f42a9af9.png)
![изображение](https://user-images.githubusercontent.com/26799424/173032230-094ccffa-eb40-41e1-94a6-547e549a6de8.png)

`default: null` тогда сделать лучше?

wspwebben replied:
Да, я ошибся: null и undefined в типе указывать нельзя
Оба твоих предложения выше: убрать required или написать валидатор подходят

efiand replied:
А `default: null` нам подойдет в таком случае вместо валидатора? Я проверил, ошибки в программе нет.

wspwebben replied:
Да, `default: null` будет лучшим вариантом на мой взгляд, я его в другом комменте и предлагал

efiand replied:
Спасибо! Видимо, я его как-то неправильно осознал.

efiand committed:
Убрано дефолтное значение user
